k2/csrc/cuda/debug.h

```diff
@@ -107,30 +111,40 @@ __host__ __device__ __forceinline__ cudaError_t _K2CudaDebug(
            threadIdx.x, threadIdx.y, threadIdx.z,
            filename, line,
            cudaGetErrorString(error));
+    if (abort) {
+      __threadfence();         // ensure memory write before trap
+      /**
+       * kill kernel (all threads) with error.
+       * It may cause context destructed.
+       * `assert(cudaSuccess != error)`
+       * is another candidate.
+       */
+      asm("trap;");
+    }
   #endif
   }
   return error;
 }
 
 /**
- * @def K2_CUDA_CHECK_ERROR(cudaError)
+ * @def K2_CUDA_PRINT_ERROR(cudaError)
  *
  * @brief Macro for checking cuda error.
  *
  * @details
- *  If error is not cudaSuccess, print the error message and exit with the
- *  error enum value. Otherwise, it does nothing.
+ *  If error is not cudaSuccess, print the error message, and pass the
+ *  optional `bAbort` as `abort` of `K2CudaDebug_`.
+ *  Otherwise, it does nothing except return the error.
  *
- * @param[in] e an enum type indicating CUDA errors.
+ * @param[in] e one in the enum type that indicates the CUDA error.
+ * @return      the CUDA error returned by `K2CudaDebug_`.
  *
  * @code{.cpp}
- * K2_CUDA_CHECK_ERROR(cudaGetLastError());
+ * K2_CUDA_PRINT_ERROR(error = cudaGetLastError());
  * @endcode
  */
-#define K2_CUDA_CHECK_ERROR(e)                                    \
-  if (::k2::_K2CudaDebug((cudaError_t)(e), __FILE__, __LINE__)) { \
-    exit(e);                                                      \
-  }
+#define K2_CUDA_PRINT_ERROR(e, bAbort...)                            \
```

Comment:
In the normal case we want to check the error and die if the status is not zero.
So please leave it as K2_CUDA_CHECK_ERROR... later if we need we can create another macro that just prints and continues.

Reply:
Done.